### PR TITLE
Refactor `websocket/` from CommonJS to ESM

### PR DESCRIPTION
**Pull request by DuCanhGH**
# Description
This PR migrates `websocket/` from CommonJS to ESM. This helps modernize the codebase, gets rid of legacy or unmaintained code (including packages), and helps make the code easier to work with. It also adds a lockfile for `npm` to avoid breaking the site when sudden breaking changes happen in one or more dependencies.

## What
Some of the changes are as aforementioned. 

The PR gets rid of `qu` (with a copy of its code) and `simplesets` (with JS's standardized `Set`), which haven't been maintained in years. 

I've also added `prettier` to help format JS code so that we don't have to manually format them (currently applies to `websocket/` only), fixed some grammar errors in `README.md` and `502.html`.

Node.js has been upgraded to v18 as v14 reached its End-of-Life in April 2023 and v16 reaches its End-of-Life today.

No breaking changes are expected from this PR.

# How was this PR tested?
I simply set `EVENT_DAEMON_USE` in `local_settings.py` to `True` and provided `EVENT_DAEMON_POST` and `EVENT_DAEMON_GET`, and verified that the websocket still worked by submitting in one tab and opening `/submissions` in another tab. Both tabs were updated when the judge finished testing.

# Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have explained the purpose of this PR.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the README/documentation
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] Informed of breaking changes, testing and migrations (if applicable).
- [ ] Attached screenshots (if applicable).

By submitting this pull request, I confirm that my contribution is made under the terms of the AGPL-3.0 License.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,6 +7,7 @@
 *.mo
 *~
 *.swp
+dmojsite/
```
Comment (hieplpvip) [inline .gitignore:10]: Why ignore `dmojsite`?

I suppose `dmojsite` is the virtual environment, but I don't think anyone would create it in the same directory as the codebase.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -44,11 +44,11 @@ jobs:
...
-      - name: Set up Node 14
-        uses: actions/setup-node@v2
+      - name: Install Node.js
+        uses: actions/setup-node@v3
         with:
-          node-version: '14'
+          node-version: '18.x'
       - name: Install npm packages
-        run: npm install -g sass postcss-cli postcss autoprefixer
+        run: npm install
```
Comment (hieplpvip) [inline .github/workflows/build.yml:52]: You may want to change this to `npm ci`

**Comment by hieplpvip** [comment]
Thanks for the PR!

I think the changes in `README.md` and `502.html` should be in another PR.
@leduythuccs What do you think?

File: .gitignore [hunk 3]
```diff
@@ -33,11 +34,9 @@ language_voj.json
...
 custom_script_*
-websocket/config.js
 node_modules/
-package-lock.json
 db.json
 tmp.sh
 *.sql
 fc_*
-local_urls.py
+local_urls.py
```
Comment (hieplpvip) [inline .gitignore:42]: Please add back the blank line

File: .github/workflows/build.yml (hunk 2 above)
Comment (DuCanhGH) [reply .github/workflows/build.yml]: @hieplpvip done in https://github.com/VNOI-Admin/OJ/pull/324/commits/3d7f306a319001e87325907450907bdb2667b628. I'm not really familar with `npm` as I usually use `pnpm`, so yeah.

File: .gitignore (hunk 1 above)
Comment (DuCanhGH) [reply .gitignore]: @hieplpvip yeah, let's just keep it local.

File: .gitignore (hunk 3 above)
Comment (DuCanhGH) [reply .gitignore]: @hieplpvip I tried in https://github.com/VNOI-Admin/OJ/pull/324/commits/3d7f306a319001e87325907450907bdb2667b628, but not sure if it's working.

**Comment by DuCanhGH** [comment]
@hieplpvip I was going to add something to `README.md` anyway, so I think it'd be nice to just keep the changes.

**Review by hieplpvip — APPROVED** [review]
LGTM

**Pull request merged**